Step1 pr 올립니다!

jinyounglee opened the pull request.

jinyounglee committed:
feat: step1 구현

Commit:
refactor: 메서드 분리, 테스트 코드 작성

nooose commented on src/main/java/nextstep/qna/validate/QnAValidator.java:
```diff
@@ -0,0 +1,19 @@
+        if (!Objects.equals(question.getWriter(), nsUser)) {
+            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
+        }
+    }
```
사용되지 않는 Validator네요 제거해도 될 것 같습니다.

nooose commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -81,12 +87,31 @@ public boolean isDeleted() {
     }
+
+    public void delete(NsUser loginUser) throws CannotDeleteException {
+        if (!Objects.equals(loginUser, writer)) {
```
loginUser와 writer가 같은지 비교하기 위해서 equals 메서드를 사용하셨네요
NsUser 간 객체 비교를 정확히 하기 위해서는 equals 메서드를 재정의 해야합니다.
그렇지 않으면 내부 멤버변수가 모두 같더라도 참조가 같은지 비교하기 때문에 결과가 다르게 나올거에요

또한 `Objects.equals(loginUser, writer)` 부분도 메서드로 만들어 볼 수 있을 것 같습니다.

nooose commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,38 @@
+    public void deleteAnswers(NsUser loginUser) throws CannotDeleteException {
+        for (Answer answer : answers) {
+            answer.validateAnswer(loginUser);
+            answer.delete();
```
answer delete 내부에서 `validateAnswer`를 호출하면 어떨까요?
`delete`가 public으로 노출되어있어서 그냥 호출될 수 있을 것 같아요
```java
answer.delete(loginUser);
```

nooose commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -26,24 +28,26 @@ public class QnAService {
+        addDeleteHistories(deleteHistories);
     }
+
+    @Transactional
+    public void deleteAnswers(Answers answers, NsUser loginUser) throws CannotDeleteException {
+        answers.deleteAnswers(loginUser);
+    }
```
상위 deleteQuestion 메서드에서  `@Transactional`이 전파되기 때문에 
이 메서드에 `@Transactional` 애너테이션은 불필요합니다.

또한 `answers.deleteAnswers(loginUser);` 자체로 의미를 가지기 때문에 메서드 추출은 굳이 안 해도 될 것 같아요

jinyounglee replied:
그렇네요..수정해 놓겠습니다.

nooose commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -26,24 +28,26 @@ public class QnAService {
+        deleteAnswers(answers, loginUser);
+
+        deleteHistories.addAllDeleteHistories(answers.deleteAllHistories());
+        addDeleteHistories(deleteHistories);
```
```java
DeleteHistories deleteHistories = question.delete(loginUser);
deleteHistoryService.saveAll(deleteHistories.getDeleteHistories());
```
우리가 원하는 것은 질문을 삭제했을 때 질문 삭제 히스토리와, 질문에 있는 답변 삭제 히스토리입니다.
delete 메서드 내부에서 삭제 히스토리 일급컬렉션을 그대로 반환할 수 있어보입니다.

서비스 레이어에서 도메인을 호출하되 도메인 안에 있는 값을 꺼내가면서 로직을 풀어나가는 방식 보다는
Question 도메인 자체적으로 해결하고 그 이후 외부 서비스를 호출하는 로직은 서비스에서 담당하면 어떨까요?

nooose commented on src/test/java/nextstep/qna/domain/QuestionTest.java:
```diff
@@ -1,8 +1,61 @@
+        내가_쓴_답변만_있는_질문.delete(NsUserTest.SANJIGI);
+        assertThat(내가_쓴_답변만_있는_질문.getAnswers().isAllDeleted()).isTrue();
+    }
+}
```
<img width="69" alt="image" src="https://github.com/next-step/java-lms/assets/36617783/950b919e-ba4f-46d7-aac1-64e1bac77b36">

이런 빨간 경고는 idea 에서 쉽게 처리할 수 있습니다.
`github eof`를 구글에 검색해 보시면 자세한 내용을 찾을 수 있을거에요

<img width="713" alt="image" src="https://github.com/next-step/java-lms/assets/36617783/3e96a5c1-4e5b-4401-87e4-5653bd6ef8b4">

nooose reviewed (changes requested):
안녕하세요 진영님!
수강신청 미션 리뷰어 성준혁입니다.

코멘트 남겼으니 반영해 보시고 재요청 부탁드립니다. 👍

Commit:
refactor: 리뷰 요청사항 반영
- 사용하지 않는 validate 클래스 제거
- Question 삭제시 사용자 비교 메서드 추출 및 equals 재정의
- validateAnswer 를 delete 메서드 내로 이동
- 불필요한 @Transactional 어노테이션 제거
- Question 삭제시 History 처리 부분 domain 영역으로 이동

Commit:
refactor: 리뷰 요청사항 반영
- 사용하지 않는 validate 클래스 제거
- Question 삭제시 사용자 비교 메서드 추출 및 equals 재정의
- validateAnswer 를 delete 메서드 내로 이동
- 불필요한 @Transactional 어노테이션 제거
- Question 삭제시 History 처리 부분 domain 영역으로 이동
